src/sparse_blas/backends/mkl_common/mkl_helper.hpp

```diff
@@ -1,56 +1,102 @@
+#include <sycl/sycl.hpp>
+#else
+#include <CL/sycl.hpp>
+#endif
 
+#include "oneapi/mkl/exceptions.hpp"
 #include "oneapi/mkl/sparse_blas/detail/helper_types.hpp"
 
+#include "sparse_blas/enum_data_types.hpp"
+#include "sparse_blas/macros.hpp"
+
 namespace oneapi::mkl::sparse::detail {
 
-inline auto get_handle(detail::matrix_handle **handle) {
-    return reinterpret_cast<oneapi::mkl::sparse::matrix_handle_t *>(handle);
+/// Return whether a pointer is accessible on the host
+template <typename T>
+inline bool is_ptr_accessible_on_host(sycl::queue &queue, const T *host_or_device_ptr) {
+    auto alloc_type = sycl::get_pointer_type(host_or_device_ptr, queue.get_context());
+    // Note sycl::usm::alloc::host may not be accessible on the host according to SYCL specification.
+    // sycl::usm::alloc::unknown is returned if the pointer is not a USM allocation which is assumed to be a normal host pointer.
+    return alloc_type == sycl::usm::alloc::shared || alloc_type == sycl::usm::alloc::unknown;
+}
+
+/// Throw an exception if the scalar is not accessible in the host
+template <typename T>
+void check_ptr_is_host_accessible(const std::string &function_name, const std::string &scalar_name,
+                                  sycl::queue &queue, const T *host_or_device_ptr) {
+    if (!is_ptr_accessible_on_host(queue, host_or_device_ptr)) {
+        throw mkl::invalid_argument(
+            "sparse_blas", function_name,
+            "Scalar " + scalar_name + " must be accessible on the host for buffer functions.");
+    }
 }
 
-inline auto get_handle(detail::matrix_handle *handle) {
-    return reinterpret_cast<oneapi::mkl::sparse::matrix_handle_t>(handle);
+/// Return a scalar on the host from a pointer to host or device memory
+/// Used for USM functions
+template <typename T>
+inline T get_scalar(sycl::queue &queue, const T *host_or_device_ptr) {
+    if (is_ptr_accessible_on_host(queue, host_or_device_ptr)) {
+        return *host_or_device_ptr;
+    }
+    T scalar;
+    auto event = queue.copy(host_or_device_ptr, &scalar, 1);
+    event.wait_and_throw();
+    return scalar;
 }
 
+/// Merge multiple event dependencies into one
+inline sycl::event collapse_dependencies(sycl::queue &queue,
+                                         const std::vector<sycl::event> &dependencies) {
+    if (dependencies.empty()) {
+        return {};
+    }
+    else if (dependencies.size() == 1) {
+        return dependencies[0];
+    }
+
+    return queue.submit([&](sycl::handler &cgh) {
+        cgh.depends_on(dependencies);
+        cgh.host_task([=]() {});
+    });
+}
+
+/// Convert \p value_type to template type argument and use it to call \p op_functor.
+#define DISPATCH_MKL_OPERATION(function_name, value_type, op_functor, ...)                         \
+    switch (value_type) {                                                                          \
+        case detail::data_type::real_fp32: return op_functor<float>(__VA_ARGS__);                  \
```

Comment:
Don't we need similar macros for APIs that have both integers and floating-point types in them?